Fix OAuth email sync to user.emails array

gauravarora opened the pull request:
## Summary
This PR fixes issue #13929 where Accounts.findUserByEmail() and other email-related APIs don't work for users who sign up via OAuth providers.

## Problem
When users authenticate via OAuth, their email is stored in service-specific fields (e.g., services.google.email) but not in the standard user.emails array. This breaks Accounts.findUserByEmail() since it searches in the emails.address field.

## Solution
Modified updateOrCreateUserFromExternalService to automatically sync email addresses from OAuth service data to the standard user.emails array for both new and existing users.

## Changes
- Modified packages/accounts-base/accounts_server.js - updateOrCreateUserFromExternalService method
- Added comprehensive test in packages/accounts-base/accounts_tests.js

Fixes #13929

gauravarora committed:
Fix OAuth email sync to user.emails array

This change ensures that when users sign up or log in via OAuth providers
(like Google, Facebook, etc.), their email address is automatically synced
to the standard user.emails array in addition to being stored in the
service-specific location (e.g., services.google.email).

Changes:
- Modified updateOrCreateUserFromExternalService to sync emails from OAuth
  service data to user.emails array for both new and existing users
- Email verification status is preserved from OAuth providers
- Prevents duplicate emails when users log in multiple times
- Maintains backward compatibility with OAuth providers that don't provide
  emails (e.g., Twitter)

This fix ensures Accounts.findUserByEmail() and other email-related APIs
work consistently regardless of whether users signed up via OAuth or
email/password.

Fixes #13929

StorytellerCZ commented on packages/accounts-base/accounts_server.js:
```diff
@@ -1401,6 +1401,28 @@ export class AccountsServer extends AccountsCommon {
+      // Sync email from OAuth service data to user.emails array if present
+      if (serviceData.email) {
+        const emailAddress = serviceData.email;
+        const emailVerified = serviceData.verified_email || serviceData.verified || false;
```
This is a tough one. We would have to check what is the general practice. I would say that OAuth and most functionalities on these platform require email verification. At least in my apps I assume that they are verified by their very nature.

gauravarora replied:
I would agree, I think it is safe to assume that the email is verified.

StorytellerCZ replied:
Also often services won't even provide a value on email verification, so the last default should be `true`.

gauravarora replied:
I removed the conditionals entirely and just pass true instead in the latest version.

StorytellerCZ commented on packages/accounts-base/accounts_server.js:
```diff
@@ -1401,6 +1401,28 @@ export class AccountsServer extends AccountsCommon {
+        const emailVerified = serviceData.verified_email || serviceData.verified || false;
+
+        // Check if the email already exists in user.emails
+        const existingUser = await this.users.findOneAsync(user._id, {
```
I would prefer a way where we don't have to have a DB call. Can we get this information from an existing source or I think the update on line 1426 could be adjusted to have a proper update that would add to the array if it exists or create it.

gauravarora replied:
Removed the call.

gauravarora committed:
Use $addToSet instead of extra DB call for email sync

Replace the separate findOneAsync call with MongoDB's $addToSet operator
to add emails to the user.emails array more efficiently. This operator
automatically handles duplicate prevention and array creation in a single
atomic operation.

gauravarora committed:
Always mark OAuth emails as verified

OAuth provider emails are assumed to be verified since the user
authenticated directly with the provider.

dupontbertrand commented:
@gauravarora @StorytellerCZ  I don’t think relying on $addToSet for emails is ideal.
The same email with verified:false and later verified:true is treated as different (the whole object is compared), which can create duplicates or trigger a uniqueness error (E11000) if there’s a unique index on emails.address
Maybe check if the address exists (case-normalize), $push if it doesn’t, otherwise $set verified:true ?

gauravarora committed:
Fix OAuth email sync to prevent duplicates and update verified flag

Replace  with logic that:
- Checks if email exists (case-insensitive)
- Uses  if email doesn't exist
- Uses  to update verified flag if email exists

This prevents duplicate emails with different verified states and avoids E11000 uniqueness errors.

gauravarora commented:
that makes sense, @StorytellerCZ i have made changes like @dupontbertrand suggested, LMK if not ok

italojs commented on packages/accounts-base/accounts_server.js:
```diff
@@ -1401,9 +1401,30 @@ export class AccountsServer extends AccountsCommon {
+          };
+        } else {
+          // Email exists, update verified to true
+          updateQuery.$set[`emails.${existingEmailIndex}.verified`] = true;
```
let me know if i'm wrong:
- I create my account
- Dont verify my email
- Create an account with Oauth using the same email as before
- Get my account verified

The verification process should occur in in [verifyEmail](https://github.com/meteor/meteor/blob/devel/packages/accounts-password/password_client.js#L266), I recommend you dont touch the updateQuery if the email already is there

StorytellerCZ replied:
So did a bit of checking. In case you have not verified your email many providers will either set the verification flag to false (which we have covered) or they will not provide email at all (in which case this flow won't happen). Many providers gate OAuth after verification to prevent the issue you are describing.
So in the end defaulting to `false` would be a defensive stance against those providers that don't follow industry standards.

italojs commented:
will you continue working in this pr @gauravarora ? you are almost there, let us know if you need any help

Grubba27 commented:
We can reopen this if you want to continue the work!
